zebra-consensus/src/block/subsidy.rs

```diff
@@ -0,0 +1,228 @@
+//! Calculation of Block Subsidy, Funding Streams, and Founders’ Reward - [§7.7][7.7]
+//!
+//! [7.7]: https://zips.z.cash/protocol/protocol.pdf#subsidies
+
+use std::convert::TryFrom;
+
+use zebra_chain::{
+    amount::{Amount, Error, NonNegative},
+    block::Height,
+    parameters::{Network, NetworkUpgrade::*},
+};
+
+use crate::parameters::subsidy::*;
+
+/// `SlowStartShift()` as described in [protocol specification §7.7][7.7]
+///
+/// [7.7]: https://zips.z.cash/protocol/protocol.pdf#subsidies
+fn slow_start_shift() -> Height {
+    Height(SLOW_START_INTERVAL.0 / 2)
+}
+
+/// `Halving(height)` as described in [protocol specification §7.7][7.7]
+///
+/// [7.7]: https://zips.z.cash/protocol/protocol.pdf#subsidies
+fn halving_factor(height: Height, network: Network) -> Result<Amount<NonNegative>, Error> {
+    if height < SLOW_START_INTERVAL {
+        panic!("can't verify before block {}", SLOW_START_INTERVAL.0)
+    }
+    let blossom_height = Blossom
+        .activation_height(network)
+        .expect("blossom activation height should be available");
+    if height >= blossom_height {
+        let scaled_halvings = (blossom_height - slow_start_shift()) as u32
+            * BLOSSOM_POW_TARGET_SPACING_RATIO as u32
+            + height.0
+            - blossom_height.0;
+        Amount::try_from((1 << scaled_halvings / POST_BLOSSOM_HALVING_INTERVAL.0) as i64)
+    } else {
+        Amount::try_from(
+            ((height - slow_start_shift()) as u32 / PRE_BLOSSOM_HALVING_INTERVAL.0) as i64,
+        )
+    }
+}
+
+/// `BlockSubsidy(height)` as described in [protocol specification §7.7][7.7]
+///
+/// [7.7]: https://zips.z.cash/protocol/protocol.pdf#subsidies
+pub fn block_subsidy(height: Height, network: Network) -> Result<Amount<NonNegative>, Error> {
+    if height < SLOW_START_INTERVAL {
+        panic!("can't verify before block {}", SLOW_START_INTERVAL.0)
+    }
+    let blossom_height = Blossom
+        .activation_height(network)
+        .expect("blossom activation height should be available");
+
+    let hf = halving_factor(height, network)?;
+    if height >= blossom_height {
+        MAX_BLOCK_SUBSIDY / BLOSSOM_POW_TARGET_SPACING_RATIO * hf
+    } else {
+        MAX_BLOCK_SUBSIDY >> hf
+    }
+}
+
+/// `FoundersReward(height)` as described in [protocol specification §7.7][7.7]
+///
+/// [7.7]: https://zips.z.cash/protocol/protocol.pdf#subsidies
+fn founders_reward(height: Height, network: Network) -> Result<Amount<NonNegative>, Error> {
+    let hf = halving_factor(height, network);
+    if hf == Amount::try_from(0) || hf == Amount::try_from(1) {
+        block_subsidy(height, network)? / FOUNDERS_FRACTION_DIVISOR
+    } else {
+        Amount::try_from(0)
+    }
```

Comment:
If you'd like, we can check for `height < Canopy` instead?

```suggestion
    if halving_divisor(height, network) == 1 {
        block_subsidy(height, network)? / FOUNDERS_FRACTION_DIVISOR
    } else {
        Amount::try_from(0)
    }
```